libexec/rtld-cheri-elf/mips/rtld_start.S

```diff
@@ -87,24 +81,6 @@ LEAF(rtld_start)
 	PCREL_LOAD_CODE_PTR($c12, t9, _rtld_relocate_nonplt_self)
 	cjalr $c12, $c17; /* (&_DYNAMIC, relocbase) */
 	cincoffset	$c3, $c4, a0	/* dynp (in delay slot) */
-	/* Returns DT_MIPS_CHERI_FLAGS value in $v0 */
-
-	/*
-	 * Now that we have processed the self relocations we can do the
-	 * __cap_relocs processing (the location needs to be adjusted to the
-	 * real address).
-	 * TODO: since there should be no external symbols in __cap_relocs we
-	 * should be able to do this before _rtld_relocate_nonplt_self and just
-	 * add the relocabase value to all targets and sources instead of the
-	 * more expensive ELF_REL processing.
-	 */
-	PCREL_LOAD_CODE_PTR($c3, a0, __start___cap_relocs)
-	PCREL_LOAD_CODE_PTR($c4, a1, __stop___cap_relocs)
-	clc		$c5, zero, SFRAME_RELOCBASE($c11) /* relocbase */
-
-	PCREL_LOAD_CODE_PTR($c12, t9, _rtld_do___caprelocs_self)
```

Comment:
Does this really work for MIPS? We need __cap_relocs processed before calling any function. It will be fine on RISCV if we use pc-relative calls, but I can't see this working in the MIPS ABIs (both PLT and PCREL).

Reply:
RISC-V and MIPS are the same, both make calls by getting capabilities from the captable, and both get them from PC-relative loads in the PCREL ABI. All I have done is inline `_rtld_do___caprelocs_self` into `_rtld_relocate_nonplt_self`. I have not introduced any new function calls, nor removed any relocation processing that happens before the instructions after this.

This is done as it's where the code belongs, as __cap_relocs are still relocations, just a different format. I could have instead made `_rtld_do___caprelocs_self` an always_inline function in cheri_relocs.h and called that, rather than manually inlining, but did _not_ do that because the code has MD bits (namely, the flags check).

I built and booted a mips-purecap world.

Reply:
Sorry, I misread the diff. Looks like jenkins is also happily running cheritest.